v2/tools/generator/internal/codegen/pipeline/apply_export_filters.go

```diff
@@ -32,47 +31,69 @@ func filterTypes(
 	configuration *config.Configuration,
 	state *State) (*State, error) {
 
-	newDefinitions := make(astmodel.Types)
-
-	filterer := configuration.BuildExportFilterer(state.types)
+	defs := make(astmodel.Types)
+	types := state.Types()
 	renames := make(map[astmodel.TypeName]astmodel.TypeName)
-
-	for _, def := range state.types {
+	for _, def := range astmodel.FindResourceTypes(types) {
 		defName := def.Name()
-		shouldExport, newName, reason := filterer(defName)
-
-		switch shouldExport {
-		case config.Skip:
-			klog.V(3).Infof("Skipping %s because %s", defName, reason)
 
-		case config.Export:
-			if reason == "" {
-				klog.V(3).Infof("Exporting %s", defName)
-			} else {
-				klog.V(2).Infof("Exporting %s because %s", defName, reason)
+		export, err := configuration.ObjectModelConfiguration.LookupExport(defName)
+		if err != nil {
+			if config.IsNotConfiguredError(err) {
+				export = false
 			}
+		}
+
+		if _, err = configuration.ObjectModelConfiguration.LookupExportAs(defName); err == nil {
+			// $exportAs is configured, we must export
+			export = true
+		}
 
-			newDefinitions[def.Name()] = def
-			if newName != nil {
-				renames[defName] = *newName
+		if !export {
+			klog.V(3).Infof("Skipping resource %s", defName)
+			continue
+		}
+
+		klog.V(3).Infof("Exporting resource %s and related types", defName)
+		typesToExport := findReferencedTypes(defName, types)
+		for n := range typesToExport {
+			defs[n] = types[n]
+
+			if as, err := configuration.ObjectModelConfiguration.LookupExportAs(n); err == nil {
+				renames[n] = n.WithName(as)
 			}
-		default:
-			panic(fmt.Sprintf("unhandled shouldExport case %q", shouldExport))
 		}
 	}
 
-	// Now apply all the renames
-	renamingVisitor := astmodel.NewRenamingVisitor(renames)
-	result, err := renamingVisitor.RenameAll(newDefinitions)
-	if err != nil {
+	if err := configuration.ObjectModelConfiguration.VerifyExportConsumed(); err != nil {
 		return nil, err
 	}
 
-	// Ensure that the export filters had no issues
-	err = configuration.GetExportFiltersError()
+	if err := configuration.ObjectModelConfiguration.VerifyExportAsConsumed(); err != nil {
+		return nil, err
+	}
+
+	// Now apply all the renames
+	renamingVisitor := astmodel.NewRenamingVisitor(renames)
+	result, err := renamingVisitor.RenameAll(defs)
 	if err != nil {
 		return nil, err
 	}
 
 	return state.WithTypes(result), nil
 }
+
+func findReferencedTypes(root astmodel.TypeName, types astmodel.Types) astmodel.TypeNameSet {
```

Comment:
You can already do this in two ways:
1. `astmodel.FindConnectedTypes` returns an `astmodel.Types` collection
2. `reference_graph.go` `graph.Connected()` returns a collection of `astmodel.TypeName`'s

IMO should use one of the existing methods

Reply:
This method was just relocated from the (now deleted ExportFilters); happy to simplify.

Reply:
Hah! I missed that it was just relocated. Yeah give one of those other methods a try. I think they should basically just work for you out of the box.